### PR TITLE
Add DataSource.remote as option

### PR DESCRIPTION
**Pull request by evanphx**
This adds DataSource.remote to go along with local and git. Remote indicates that server has some bespoke way to fetch the data source. The server in question can use the populateDataSource hook to convert the DataSource into another format before the job is queued.

### CODE REVIEW

File: pkg/server/singleprocess/service_job.go [hunk 1]
```diff
@@ -133,6 +133,33 @@ func (s *Service) queueJobReqToJob(
...
 	}
 
+	// If the data source is set to remote, then let the server hook populate it.
+	if job.DataSource.GetRemote() != nil {
+		if s.populateDataSource != nil {
+			job, err = s.populateDataSource(ctx, job)
+			if err != nil {
+				log.Error("error populating data source for job", "error", err)
+				return nil, "", status.Errorf(codes.Internal,
+					"An internal server issue was detected when calculate the data source")
```
Comment (jgwhite) [inline pkg/server/singleprocess/service_job.go:143]:
```suggestion
					"An internal server issue was detected when calculating the data source")
```

File: pkg/server/singleprocess/service_job.go [hunk 2]
```diff
@@ -133,6 +133,33 @@ func (s *Service) queueJobReqToJob(
...
+			}
+
+			// The new job can't still have remote, so if it wasn't updated, then
+			// error out.
+			if job.DataSource.GetRemote() != nil {
+				log.Error("populateDataSource returned another remote DS job")
+				return nil, "", status.Errorf(codes.Internal,
+					"An internal server issue was detected when calculate the data source")
```
Comment (jgwhite) [inline pkg/server/singleprocess/service_job.go:151]:
```suggestion
					"An internal server issue was detected when calculating the data source")
```

File: pkg/server/singleprocess/service_job.go [hunk 3]
```diff
@@ -133,6 +133,33 @@ func (s *Service) queueJobReqToJob(
...
+			}
+		} else {
+			log.Error("job has a remote DataSource but server provided to populateDataSource")
+			// This is a server misconfiguration.
+			if job.DataSource.GetRemote() != nil {
+				return nil, "", status.Errorf(codes.Internal,
+					"An internal server issue was detected when calculate the data source")
```
Comment (jgwhite) [inline pkg/server/singleprocess/service_job.go:158]:
```suggestion
					"An internal server issue was detected when calculating the data source")
```

**Review by jgwhite — APPROVED** [review]
Looks good from my perspective. Spotted a few typos in log lines, but no other feedback.

**Review by briancain — APPROVED** [review]
Approved but +1 on Jamies doc string fixes!

**Review by izaaklauer — COMMENTED** [review]
👍

**Pull request merged**